[Swift in WebKit] Work towards modularizing PAL (part 2)

rr-codes opened the pull request:
#### f6491e12ca0be45502268e4cb24a105e0d274d76
<pre>
[Swift in WebKit] Work towards modularizing PAL (part 2)
<a href="https://bugs.webkit.org/show_bug.cgi?id=303250">https://bugs.webkit.org/show_bug.cgi?id=303250</a>
<a href="https://rdar.apple.com/165544156">rdar://165544156</a>

Reviewed by Mike Wyrzykowski.

More fixes made to previously incorrect PAL code.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:

- Remove PALSwiftUtils.h, and add CryptoDigestHashFunction.h (as a private-level header).

* Source/WebCore/PAL/pal/PALSwift.h:

- Remove unused `OptionalVectorUInt8` definition
- Because of the way PAL is currently incorrectly structured for Swift, an include of `CryptoDigestHashFunction.h` is needed before the generated Swift header is included.

* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
* Source/WebCore/PAL/pal/crypto/CryptoDigest.h:
* Source/WebCore/PAL/pal/crypto/CryptoDigestHashFunction.h:

Previously, both CryptoKitShim.swift and CryptoDigest.h had more or less the same type to represent a crypto digest hash function. De-duplicate this
logic by creating a separate C++ type shared by both files. This enum is C++ instead of Swift because (a) Swift-Cxx interop is significantly better
supported compared to reverse-interop, and (b) it is easier for this type to be used by a seperate module (aka WebCore) since it is invalid
for C++/Objective C to call Swift code from a different framework/module.

* Source/WebCore/PAL/pal/PALSwiftUtils.h: Removed.

Delete this file entirely, for several reasons:

- It doesn&apos;t really have anything to do with Swift specifically in the first place
- It defines symbols inside the WebCore namespace albeit being in PAL
- It was a project header included from a different framework, which is invalid
- It was dependent on the generated Swift header, which cannot cross module/framework boundaries
- It created a layering violation and dependency cycle since the file uses `CryptoAlgorithmIdentifier`, a type defined in WebCore
- It only &quot;worked&quot; in the first place because the functions inside of it get inlined into WebCore and was always an invalid header file
- The functions inside it are only used in WebCore

* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
(WebCore::toCKHashFunction):
(WebCore::isValidHashParameter):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:

- Move the functions from PALSwiftUtils to CryptoAlgorithmIdentifier

* Source/WebCore/PAL/pal/module.modulemap:

- Temporarily add a sub-module for Swift to be able to access `CryptoDigestHashFunction`; soon, this module map will be fixed to use an umbrella
header, so this will not be needed then.

* Source/WebCore/PAL/pal/spi/cocoa/AVStreamDataParserSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/NSAccessibilitySPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/URLFormattingSPI.h:
* Source/WebCore/PAL/pal/spi/ios/MediaPlayerSPI.h:
* Source/WebCore/PAL/pal/spi/ios/OpenGLESSPI.h:
* Source/WebCore/PAL/pal/spi/ios/QuickLookSPI.h:
* Source/WebCore/PAL/pal/spi/ios/SBSStatusBarSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSImmediateActionGestureRecognizerSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPasteboardSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPopoverColorWellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPopoverSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSScrollingMomentumCalculatorSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSearchFieldCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServicePickerSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServiceSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextFieldCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextTableSPI.h:
* Source/WebKit/UIProcess/mac/WebColorPickerMac.mm:
(-[WKPopoverColorWell _showPopover]):

- Add missing compile time guards
- Add FIXMEs for issues that prevent proper modularization of PAL

Canonical link: <a href="https://commits.webkit.org/303648@main">https://commits.webkit.org/303648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38c4111212690315bbb949689f113074fd55c3a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133165 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140718 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85205 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/732318e5-a0fa-4cc6-8f9e-a2fbdf2e05cf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135035 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6169 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101850 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69276 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f66656a0-0340-47d9-b18f-327b8fef8369) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136112 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119349 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82646 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/379ecb75-16d1-4894-8fd5-848a7aea1da8) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4256 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1832 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113333 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37467 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143367 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5336 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38044 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110228 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5418 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4588 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110410 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27977 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4129 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115606 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59056 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5391 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33960 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5237 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68843 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5480 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5347 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
✅ with build errors addressed

rr-codes committed:
[Swift in WebKit] Work towards modularizing PAL (part 2)
https://bugs.webkit.org/show_bug.cgi?id=303250
rdar://165544156

Reviewed by Mike Wyrzykowski.

More fixes made to previously incorrect PAL code.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:

- Remove PALSwiftUtils.h, and add CryptoDigestHashFunction.h (as a private-level header).

* Source/WebCore/PAL/pal/PALSwift.h:

- Remove unused `OptionalVectorUInt8` definition
- Because of the way PAL is currently incorrectly structured for Swift, an include of `CryptoDigestHashFunction.h` is needed before the generated Swift header is included.

* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
* Source/WebCore/PAL/pal/crypto/CryptoDigest.h:
* Source/WebCore/PAL/pal/crypto/CryptoDigestHashFunction.h:

Previously, both CryptoKitShim.swift and CryptoDigest.h had more or less the same type to represent a crypto digest hash function. De-duplicate this
logic by creating a separate C++ type shared by both files. This enum is C++ instead of Swift because (a) Swift-Cxx interop is significantly better
supported compared to reverse-interop, and (b) it is easier for this type to be used by a seperate module (aka WebCore) since it is invalid
for C++/Objective C to call Swift code from a different framework/module.

* Source/WebCore/PAL/pal/PALSwiftUtils.h: Removed.

Delete this file entirely, for several reasons:

- It doesn't really have anything to do with Swift specifically in the first place
- It defines symbols inside the WebCore namespace albeit being in PAL
- It was a project header included from a different framework, which is invalid
- It was dependent on the generated Swift header, which cannot cross module/framework boundaries
- It created a layering violation and dependency cycle since the file uses `CryptoAlgorithmIdentifier`, a type defined in WebCore
- It only "worked" in the first place because the functions inside of it get inlined into WebCore and was always an invalid header file
- The functions inside it are only used in WebCore

* Source/WebCore/crypto/CryptoAlgorithmIdentifier.h:
(WebCore::toCKHashFunction):
(WebCore::isValidHashParameter):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHMACMac.cpp:

- Move the functions from PALSwiftUtils to CryptoAlgorithmIdentifier

* Source/WebCore/PAL/pal/module.modulemap:

- Temporarily add a sub-module for Swift to be able to access `CryptoDigestHashFunction`; soon, this module map will be fixed to use an umbrella
header, so this will not be needed then.

* Source/WebCore/PAL/pal/spi/cocoa/AVStreamDataParserSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/NSAccessibilitySPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/URLFormattingSPI.h:
* Source/WebCore/PAL/pal/spi/ios/MediaPlayerSPI.h:
* Source/WebCore/PAL/pal/spi/ios/OpenGLESSPI.h:
* Source/WebCore/PAL/pal/spi/ios/QuickLookSPI.h:
* Source/WebCore/PAL/pal/spi/ios/SBSStatusBarSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSImmediateActionGestureRecognizerSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPasteboardSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPopoverColorWellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSPopoverSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSScrollingMomentumCalculatorSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSearchFieldCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServicePickerSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSSharingServiceSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextFieldCellSPI.h:
* Source/WebCore/PAL/pal/spi/mac/NSTextTableSPI.h:
* Source/WebKit/UIProcess/mac/WebColorPickerMac.mm:
(-[WKPopoverColorWell _showPopover]):

- Add missing compile time guards
- Add FIXMEs for issues that prevent proper modularization of PAL

Canonical link: https://commits.webkit.org/303648@main
